src/diffTool/rulesProcessor.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {
+  Engine,
+  RuleProperties,
+  Event,
+  Almanac,
+  RuleResult,
+  EngineResult
+} from "json-rules-engine";
+import { NodeDiff } from "./jsonDiff";
+import fs from "fs-extra";
+import { ramlToolLogger } from "../logger";
+/* eslint-disable @typescript-eslint/no-use-before-define */
+
+//ID for the diff node/fact that is passed to the rule engine
+export const DIFF_FACT_ID = "diff";
+
+/**
+ * Apply rules on the AMF node differences, updates passed rules details in the node
+ * @param diffs - Array of differences
+ * @param rulesPath - File path of the rules
+ */
+export async function applyRules(
+  diffs: NodeDiff[],
+  rulesPath: string
+): Promise<void> {
+  if (!diffs || diffs.length == 0) {
+    ramlToolLogger.info("No differences to apply the rules");
+    return;
+  }
+  const rules = getRules(rulesPath);
+  if (!rules || rules.length == 0) {
+    ramlToolLogger.info("No rules to apply on the differences");
+    return;
+  }
+  //initialize engine with rules
+  const engine = new Engine(rules);
+  //callback function to execute when a rule is passed/evaluates to true
+  engine.on("success", successHandler);
+  addCustomOperators(engine);
+
+  //run rules on diffs
+  const promises = diffs.map(diff => {
+    return runEngine(engine, diff);
+  });
+  await Promise.all(promises);
+}
+
+/**
+ * Get rules from rules json file
+ * @param rulesPath - Path to rules json file
+ *
+ * @returns Array of rules
+ */
+function getRules(rulesPath: string): RuleProperties[] {
+  if (!rulesPath) {
```

Comment:
This is only testing that the rulesPath is defined, not that it exists. Should we be doing that? And if not why don't we  let readJSONSync throw if rulesPath is undefined?

Reply:
readJSONSync throws error when the file do not exist. I have a test for that
`/tmp/no-rules.json: ENOENT: no such file or directory`

Reply:
If that is the case, what purpose does this conditional server?

Reply:
Initially I was not modifying the error message in the catch and the default message was not user friendly. We don't need it anymore. I will remove it